### PR TITLE
Saved payment methods 💳 

### PR DESCRIPTION
**Pull request by danbillson**
# Saved payment methods

Adds the resources for saved payment methods - https://developer.paddle.com/api-reference/payment-methods/overview

List saved payment methods for a customer
```
  const collection = paddle.paymentMethods.list(customerId);
  const list = await collection.next();
```

Get a single saved payment method for a customer
```
const paymentMethod = await paddle.paymentMethods.get(customerId, paymentMethodId);
```

Delete a saved payment method (Cannot delete a payment method that is in use)
```
await paddle.paymentMethods.delete(customerId, paymentMethodId);
```

## Notifications

Adds new webhook notification types - https://developer.paddle.com/webhooks/payment-methods/payment-method-saved

image

## Generate auth token

Related docs: https://developer.paddle.com/api-reference/customers/generate-customer-authentication-token

```
const authToken = await paddle.customers.generateAuthToken(customerId);
```

### CODE REVIEW

**Review by davidgrayston-paddle — COMMENTED** [review]
Looks good! 👍 

I think we also need to add the notification entities and map in `src/notifications/helpers/webhooks.ts`

File: src/entities/index.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export * from './subscription';
 export * from './address';
 export * from './discount';
 export * from './events';
+export * from './payment-method';
```
Comment (mikeymike) [inline src/entities/index.ts:18]: Maybe not important but noticed an odd mix of singular and plurals here 👀

File: src/entities/payment-method/index.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/**
+ *  ! Autogenerated code !
+ *  Do not make changes to this file.
+ *  Changes may be overwritten as part of auto-generation.
+ */
+
+export * from './card';
+export * from './pay-pal';
```
Comment (mikeymike) [inline src/entities/payment-method/index.ts:8]: This doesn't quite look right to me, can see it's the file name but should we have the `-` if it's a single word/brand

File: src/types/payment-method/index.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ *  ! Autogenerated code !
+ *  Do not make changes to this file.
+ *  Changes may be overwritten as part of auto-generation.
+ */
+
+export * from './card';
+export * from './pay-pal';
```
Comment (mikeymike) [inline src/types/payment-method/index.ts:8]: +1 query on file naming 👀

**Review by mikeymike — COMMENTED** [review]
Overall looks solid, just a couple comments around naming files 😅

File: src/entities/index.ts (hunk 1 above)
Comment (danbillson) [reply src/entities/index.ts:18]: Looks like entities should all be singular, may look at doing a bit of a tidy in the v2 release

File: src/entities/payment-method/card.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/**
+ *  ! Autogenerated code !
+ *  Do not make changes to this file.
+ *  Changes may be overwritten as part of auto-generation.
+ */
+
+import type { SavedPaymentMethodCardType } from '../../enums';
+import type { ICardResponse } from '../../types';
+
+export class Card {
```
Comment (davidgrayston-paddle) [inline src/entities/payment-method/card.ts:10]: Could the existing `src/entities/shared/payment-card.ts` be used instead?

File: src/notifications/entities/payment-method/payment-method-notification.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  public readonly type: SavedPaymentMethodType;
+  public readonly origin: SavedPaymentOrigin;
+  public readonly savedAt: string;
+  public readonly updatedAt: string;
+
+  constructor(paymentMethod: IPaymentMethodNotificationResponse) {
+    this.id = paymentMethod.id;
+    this.customerId = paymentMethod.customer_id;
+    this.addressId = paymentMethod.address_id;
+    this.deletionReason = paymentMethod.deletion_reason;
```
Comment (davidgrayston-paddle) [inline src/notifications/entities/payment-method/payment-method-notification.ts:24]: Should there be a separate response type for deleted and saved? e.g.
- `IPaymentMethodNotificationResponse` (without `deletionReason`)
- `IPaymentMethodDeletedNotificationResponse` (with `deletionReason`)

Comment (danbillson) [reply src/notifications/entities/payment-method/payment-method-notification.ts:24]: Yeah, it probably is worth having the explicit type defined

File: src/entities/payment-method/card.ts (hunk 4 above)
Comment (danbillson) [reply src/entities/payment-method/card.ts:10]: It absolutely can, will clean this up

File: src/enums/payment-method/deletion-reason.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+/**
+ *  ! Autogenerated code !
+ *  Do not make changes to this file.
+ *  Changes may be overwritten as part of auto-generation.
+ */
+
+export type SavedPaymentDeletionReason = 'replaced_by_newer_version' | 'api';
```
Comment (mikeymike) [inline src/enums/payment-method/deletion-reason.ts:7]: should this be `SavedPaymentMethodDeletionReason` or `PaymentMethodDeletionReason` instead?

**Review by davidgrayston-paddle — APPROVED** [review]

**Pull request merged**